Description: Avoid race condition in test by re-trying up to 5 times
Author: Clint Byrum <clint@ubuntu.com>
Origin: http://code.google.com/p/memcached/issues/detail?id=252

Index: memcached/programs/testapp.c
===================================================================
--- memcached.orig/programs/testapp.c    2012-04-04 17:19:28.610145255 +0100
+++ memcached/programs/testapp.c    2012-04-04 17:29:03.231682461 +0100
@@ -433,8 +433,8 @@
     if (daemon) {
         /* loop and wait for the pid file.. There is a potential race
          * condition that the server just created the file but isn't
-         * finished writing the content, but I'll take the chance....
-         */
+         * finished writing the content, so we loop a few times
+         * reading as well */
         while (access(pid_file, F_OK) == -1) {
             usleep(10);
         }
@@ -445,7 +445,11 @@
                     strerror(errno));
             assert(false);
         }
-        assert(fgets(buffer, sizeof(buffer), fp) != NULL);
+
+        /* Avoid race by retrying 20 times */
+        for (int x = 0; x < 20 && fgets(buffer, sizeof(buffer), fp) == NULL; x++) {
+            usleep(10);
+        }
         fclose(fp);

         int32_t val;
